### PR TITLE
shell: Animate host switcher caret

### PR DESCRIPTION
**Pull request by MeastroZI**
"Introducing a minor update:

Changes: Improved caret transition in the document, accessible through .

View the modifications : 

https://github.com/cockpit-project/cockpit/assets/106718914/efe6945f-0bdf-42d6-bf81-e4c2112478be

Previous version  : 

https://github.com/cockpit-project/cockpit/assets/106718914/5098e445-23d8-4531-8aac-d71249965999

### CODE REVIEW

File: pkg/shell/hosts.jsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ import PropTypes from 'prop-types';
 import { PageSidebar } from "@patternfly/react-core/dist/esm/components/Page/index.js";
 import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
 import { Tooltip } from "@patternfly/react-core/dist/esm/components/Tooltip/index.js";
-import { EditIcon, MinusIcon, CaretUpIcon, CaretDownIcon } from '@patternfly/react-icons';
+import { EditIcon, MinusIcon, CaretUpIcon, /* CaretDownIcon */ } from '@patternfly/react-icons';
```
Comment (martinpitt) [inline pkg/shell/hosts.jsx:9]: Please don't comment out imports or other code, just drop them.

File: pkg/shell/hosts.jsx [hunk 2]
```diff
@@ -184,13 +184,14 @@ export class CockpitHosts extends React.Component {
                                     <HostLine user={user} host={label} />
                                 </span>
                             </span>
-                            {this.state.opened && <CaretUpIcon className="pf-v5-c-select__toggle-arrow mobile_v pf-v5-c-icon pf-m-lg" aria-hidden="true" />}
-                            {!this.state.opened && <CaretDownIcon className="pf-v5-c-select__toggle-arrow mobile_v pf-v5-c-icon pf-m-lg" aria-hidden="true" />}
+                            
+                            <CaretUpIcon className={`pf-v5-c-select__toggle-arrow mobile_v  ${this.state.opened ? 'clicked' : ''
+                            } pf-v5-c-icon pf-m-lg`} aria-hidden="true" />
```
Comment (martinpitt) [inline pkg/shell/hosts.jsx:189]: This is indented badly, the `}` shouldn't be on the same column as the initial tag, but indented more. It usually looks best to put one attribute per line, and aligning them.

File: pkg/shell/nav.scss [hunk 3]
```diff
@@ -57,7 +57,7 @@ $desktop: $phone + 1px;
     padding-block: var(--pf-v5-global--spacer--md) var(--pf-v5-global--spacer--sm);
     padding-inline: var(--pf-v5-global--spacer--md);
 
-    > .pf-v5-c-text-input-group__main {
+    >.pf-v5-c-text-input-group__main {
```
Comment (martinpitt) [inline pkg/shell/nav.scss:60]: Please revert these changes (also the ones below), they are unrelated to the caret change and just noise. In general, unrelated changes should be in separate commits.

File: pkg/shell/nav.scss [hunk 4]
```diff
@@ -171,6 +171,7 @@ $desktop: $phone + 1px;
 }
 
 [dir="rtl"] {
+
```
Comment (martinpitt) [inline pkg/shell/nav.scss:174]: Likewise, there's several spurious newlines which don't belong here. Please revert. The only relevant change for the caret is the introduction of the `.clicked` class.

File: pkg/shell/nav.scss [hunk 5]
```diff
@@ -258,6 +260,17 @@ $desktop: $phone + 1px;
...
   }
+
+  .pf-v5-c-select__toggle-arrow {
+    transition-duration: 0.3s;
+    margin: 5px;
+    font-size: 1.3rem;
+  }
+
+  .clicked {
+    transform: rotate(180deg);
```
Comment (martinpitt) [inline pkg/shell/nav.scss:271]: So how is that better than using `CaretDownIcon`? This needs some justification. We generally prefer using PF  components over custom CSS, and that introduces *more* custom CSS -- so it should have a good reason. (I'm not saying it's wrong, I just don't understand it).

Your video shows that the caret is now white on the dark gray background, whereas current main version is black on dark gray, which has poor contrast. That's good, I just don't understand what change here actually changes the color?

File: pkg/shell/nav.scss [hunk 6]
```diff
@@ -797,14 +837,44 @@ $desktop: $phone + 1px;
 /* Navigation animation */
 
 @keyframes navHostsSlide {
-  0% { opacity: 0.25; transform: scaleY(0); }
-  25% { opacity: 0.5; transform: scaleY(0.25); }
-  100% { opacity: 1; transform: scaleY(1); }
+  0% {
+    opacity: 0.25;
+    transform: scaleY(0);
+  }
```
Comment (martinpitt) [inline pkg/shell/nav.scss:843]: This is a matter of opinion. IMHO it's over-splitting, the lines are short enough that a compact paragraph is easier to understand, and more importantly, visually compare. If you have a strong opinion about this, please move that into a separate commit, otherwise revert.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks for your first contribution, exciting! This still needs some work though, I left some comments below.

File: pkg/shell/nav.scss (hunk 5 above)
Comment (MeastroZI) [reply pkg/shell/nav.scss:271]: In the previous version of the cockpit, achieving a smooth transition for the rotation of the caret from 0 to 180 degrees posed a challenge. We were using two different components—one for the caret facing down and another for the caret facing up. I addressed this by consolidating the two components into one, incorporating a transition that smoothly rotates from 0 to 180 degrees and vice versa. **The result is a more satisfying and  pleasing animation.** My intention was to streamline the code by using a single component and toggling a class to handle the rotation transformation. While it's possible to achieve this in one line within the component, I personally find the approach of adding and removing classes for toggling more readable and maintainable.

**Comment by martinpitt** [comment]
@garrett I'm afraid this isn't really my area of expertise. Could you please have a look? The change looks technically plausible now. Thanks @MeastroZI !

**Review by garrett — DISMISSED** [review]
Thanks!

We already had caret animation like this in other parts of Cockpit (expandable tables, for example). Thanks for adding it to the dropdown of the host switcher too!

**Comment by martinpitt** [comment]
Thanks @MeastroZI ! I rebased your branch, fixed the commit message and the trailing space in the CSS, and drive this to completion. I also tested it and it works nicely now. Cheers! :tada:

**Comment by martinpitt** [comment]
This [pixel test failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19815-20240123-055629-f9c581f5-fedora-39-expensive/log.html) points out a problem, the caret isn't yet animated to close again. Most likely we'll just need to wait a bit, checking.

**Comment by martinpitt** [comment]
Ah, it is wrong because the caret is really the wrong way around -- it should point down when closed, and point up when opened. I flipped it around.

The new caret is also a bit bigger, which is fine. updating the pixel tests accordingly.

**Review by martinpitt — APPROVED** [review]
